File: SecureUWPClient/SecureUWPClient/Handshake/IOMessageExhange.cs
using SecureUWPChannel.Interfaces;
using SecureUWPChannel.IOTransport;
using SecureUWPChannel.Prooperties;
using SecureUWPChannel.Serialization;
using SecureUWPClient.Ciphers.AES;
using SecureUWPClient.Ciphers.RSA;
using SecureUWPClient.Ciphers.Symmetric;
using SecureUWPClient.Configuration;
using SecureUWPClient.KeyManager;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SecureUWPClient.Handshake
{

    public class IOMessageExhange: IOCallbackAsync
    {

        private IOMulticastAndBroadcast ActivitySocket;
        private CiphersForUse ciphersforUse;
        private KeyManagerImp keystore;


        private JSonObject ReadObj;
        private JSonObject WriteObj;

        public IOMessageExhange(IOMulticastAndBroadcast ActivitySocket, CiphersForUse ciphersforUse, KeyManagerImp keystore)
        {
            this.ActivitySocket = ActivitySocket;
            this.ciphersforUse = ciphersforUse;
            this.keystore = keystore;
            Intialize();
        }

        public void Intialize()
        {
            ReadObj = new JSonObject();
            WriteObj = new JSonObject();
        }

        public override async Task SendDHEncryptedMessage(string Message, AES_CBC CBC)
        {
            WriteObj = new JSonObject();
            String encrypted = await CBC.AeS_Encrypt(Message, await keystore.LoadCipherKey());
            WriteObj.EncryptedMessage = encrypted;
            WriteObj.HmacHash = HMacAlgoProvider.CreateHMAC(encrypted, await keystore.LoadIntegrityKey(), ciphersforUse.HashAlgorithm);
            // System.out.println(ObjToSend.EncryptedMessage);

            String toSend = JsonParse.WriteObject(WriteObj);
            await this.ActivitySocket.send(toSend);
            return;
        }

        public override async Task SendDHEncryptedMessage(string Message, AES_ECB ECB)
        {
            WriteObj = new JSonObject();
            String encrypted = await ECB.AeS_Encrypt(Message, await keystore.LoadCipherKey());
            WriteObj.EncryptedMessage = encrypted;
            WriteObj.HmacHash = HMacAlgoProvider.CreateHMAC(encrypted, await keystore.LoadIntegrityKey(), ciphersforUse.HashAlgorithm);
            // System.out.println(ObjToSend.EncryptedMessage);

            String toSend = JsonParse.WriteObject(WriteObj);
            await this.ActivitySocket.send(toSend);
            return;
        }

        public override async Task<string> ReceiveDHEncryptedMessage(AES_CBC CBC)
        {
            ReadObj = JsonParse.ReadObject(await this.ActivitySocket.read());
            if (Fingerprint.VerifySignature(ReadObj.EncryptedMessage, await keystore.LoadPublicKey(), ReadObj.FingerPrint))
            {
                if (HMacAlgoProvider.VerifyHMAC(ReadObj.EncryptedMessage, await keystore.LoadIntegrityKey(), ReadObj.HmacHash,
                        ciphersforUse.HashAlgorithm))
                {
                    return await CBC.AeS_Decrypt(ReadObj.EncryptedMessage,
                           await keystore.LoadCipherKey());
                }
                else {
                    throw new Exception("Integrity of SymmetricKey canot verified");
                }
            }
            else {
                throw new Exception("Integrity of RSA canot verified");
            }
        }


        public override async Task<string> ReceiveDHEncryptedMessage(AES_ECB ECB)
        {
            ReadObj = JsonParse.ReadObject(await this.ActivitySocket.read());
            if (Fingerprint.VerifySignature(ReadObj.EncryptedMessage, await keystore.LoadPublicKey(), ReadObj.FingerPrint))
            {
                if (HMacAlgoProvider.VerifyHMAC(ReadObj.EncryptedMessage, await keystore.LoadIntegrityKey(), ReadObj.HmacHash,
                        ciphersforUse.HashAlgorithm))
                {
                    return await ECB.AeS_Decrypt(ReadObj.EncryptedMessage,
                           await keystore.LoadCipherKey());
                }
                else {
                    throw new Exception("Integrity of SymmetricKey canot verified");
                }
            }
            else {
                throw new Exception("Integrity of RSA canot verified");
            }
        }

    }
}

File: SecureAndroidClient/app/src/main/java/com/security/crypto/Handshake/IOMessageExhange.java
package com.security.crypto.Handshake;

import com.security.crypto.Ciphers.AES.AesCBC;
import com.security.crypto.Ciphers.AES.AesECB;
import com.security.crypto.Ciphers.AES.HMacAlgoProvider;
import com.security.crypto.Ciphers.RSA.Fingerprint;
import com.security.crypto.Configuration.CiphersForUse;
import com.security.crypto.Configuration.JSonObject;
import com.security.crypto.IOSocket.IOCallback;
import com.security.crypto.IOSocket.IOTransport;
import com.security.crypto.KeyManager.KeyManagerImp;


public class IOMessageExhange extends IOCallback {

    private KeyManagerImp keystore = null;
    private IOTransport SocketChanel = null;
    private CiphersForUse ciphersforUse = null;

    public IOMessageExhange(IOTransport SocketChanel, KeyManagerImp keystore, CiphersForUse ciphersforUse) {
        this.SocketChanel = SocketChanel;
        this.keystore = keystore;
        this.ciphersforUse = ciphersforUse;
    }


    @Override
    public void SendDHEncryptedMessage(String Message, AesCBC CBC) throws Exception {
        JSonObject ObjToSend = new JSonObject();
        String encrypted = CBC.AeS_Encrypt(Message, keystore.loadRemoteCipherKey());
        ObjToSend.EncryptedMessage = encrypted;
        ObjToSend.HmacHash = HMacAlgoProvider.HmacSign(encrypted, keystore.loadRemoteIntegrityKey(), ciphersforUse.getHashAlgorithm());
        // System.out.println(ObjToSend.EncryptedMessage);

        String toSend = JSonParse.WriteObject(ObjToSend);
        SocketChanel.SendMessage(toSend);
    }

    @Override
    public void SendDHEncryptedMessage(String Message, AesECB ECB) throws Exception {
        JSonObject ObjToSend = new JSonObject();
        String encrypted = ECB.AeS_Encrypt(Message, keystore.loadRemoteCipherKey());
        ObjToSend.EncryptedMessage = encrypted;
        ObjToSend.HmacHash = HMacAlgoProvider.HmacSign(encrypted, keystore.loadRemoteIntegrityKey(), ciphersforUse.getHashAlgorithm());
        //  System.out.println(ObjToSend.EncryptedMessage);

        String toSend = JSonParse.WriteObject(ObjToSend);
        SocketChanel.SendMessage(toSend);
    }

    @Override
    public String ReceiveDHEncryptedMessage(AesCBC CBC) throws Exception {
        JSonObject receivedObj = JSonParse.ReadObject(SocketChanel.receiveMessage());
        if (Fingerprint.VerifySig(receivedObj.EncryptedMessage, keystore.loadCertificate(), receivedObj.FingerPrint)) {
            if (HMacAlgoProvider.HmacVerify(receivedObj.EncryptedMessage, keystore.loadRemoteIntegrityKey(), receivedObj.HmacHash,
                    ciphersforUse.getHashAlgorithm())) {
                return CBC.AeS_Decrypt(receivedObj.EncryptedMessage,
                        keystore.loadRemoteCipherKey());
            } else {
                throw new Exception("Integrity of SymmetricKey canot verified");
            }
        } else {
            throw new Exception("Integrity of RSA canot verified");
        }

    }

    @Override
    public String ReceiveDHEncryptedMessage(AesECB ECB) throws Exception {
        JSonObject receivedObj = JSonParse.ReadObject(SocketChanel.receiveMessage());
        if (Fingerprint.VerifySig(receivedObj.EncryptedMessage, keystore.loadCertificate(), receivedObj.FingerPrint)) {
            if (HMacAlgoProvider.HmacVerify(receivedObj.EncryptedMessage, keystore.loadRemoteIntegrityKey(), receivedObj.HmacHash,
                    ciphersforUse.getHashAlgorithm())) {
                return ECB.AeS_Decrypt(receivedObj.EncryptedMessage,
                        keystore.loadRemoteCipherKey());
            } else {
                throw new Exception("Integrity of SymmetricKey cannot verified");
            }
        } else {
            throw new Exception("Integrity of RSA cannot verified");
        }

    }

    @Override
    public String toString() {
        return super.toString();
    }
}
